estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package estransport
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/url"
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	defaultResurrectTimeoutInitial      = 60 * time.Second
+	defaultResurrectTimeoutFactorCutoff = 5
+)
+
+// ConnectionPool defines the interface for the connection pool.
+//
+type ConnectionPool interface {
+	Next() (*Connection, error)
+	Remove(*Connection) error
+}
+
+// Connection represents a connection to a node.
+//
+type Connection struct {
+	sync.Mutex
+
+	URL       *url.URL
+	Dead      bool
+	DeadSince time.Time
+	Failures  int
+}
+
+type singleConnectionPool struct {
+	connection *Connection
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+type roundRobinConnectionPool struct {
+	sync.Mutex
+
+	curr int           // Index of the current connection
+	live []*Connection // List of live connections
+	dead []*Connection // List of dead connections
+	orig []*url.URL    // List of original URLs, passed in during initialization
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+func newSingleConnectionPool(u *url.URL) *singleConnectionPool {
+	cp := singleConnectionPool{connection: &Connection{URL: u}}
+
+	return &cp
+}
+
+func newRoundRobinConnectionPool(u ...*url.URL) *roundRobinConnectionPool {
+	var conns []*Connection
+	for _, url := range u {
+		conns = append(conns, &Connection{URL: url})
+	}
+
+	cp := roundRobinConnectionPool{live: conns, orig: u, curr: -1}
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.live = cp.live
+		cp.metrics.dead = cp.dead
+		cp.metrics.Unlock()
+	}
+
+	return &cp
+}
+
+// Next returns the connection from pool.
+//
+func (cp *singleConnectionPool) Next() (*Connection, error) {
+	return cp.connection, nil
+}
+
+// Remove is a no-op for single connection pool.
+//
+func (cp *singleConnectionPool) Remove(c *Connection) error {
+	return nil
+}
```

Comment:
would it make sense to have a cirtcuite breaker or some backoff that disables the connection for N seconds if it has been 'removed' ?

Reply:
That's effectively hapenning with the "scheduled resurrect" — as soon as the connection is removed, it is not being returned from `Next()`, and the scheduler will try to "resurrect" it with an exponential delay.

Reply:
Yes, the round robin one has resurrection and retry support. But if only one URL is configured we will use singleConnectionPool, which doesn't seem to have some kind of backoff behavior.